src/data/set/ncard.lean

```diff
@@ -0,0 +1,750 @@
+meta def to_finite_tac : tactic unit := `[exact set.to_finite _]
+
+lemma ncard_def (s : set α) : s.ncard = nat.card s := rfl
+
+lemma ncard_eq_to_finset_card (s : set α) (hs : s.finite . to_finite_tac) :
+  s.ncard = hs.to_finset.card :=
+by rw [ncard_def, @nat.card_eq_fintype_card _ hs.fintype,
+  @finite.card_to_finset _ _ hs.fintype hs]
+
+lemma ncard_le_of_subset (hst : s ⊆ t) (ht : t.finite . to_finite_tac) : s.ncard ≤ t.ncard :=
+@finite.card_le_of_embedding _ _ (finite_coe_iff.mpr ht) (set.embedding_of_subset _ _ hst)
+
+lemma ncard_mono [finite α] : @monotone (set α) _ _ _ ncard := λ _ _, ncard_le_of_subset
+
+@[simp] lemma ncard_eq_zero (hs : s.finite . to_finite_tac) : s.ncard = 0 ↔ s = ∅ :=
+by simp [ncard_def, @finite.card_eq_zero_iff _ hs.to_subtype]
+
+@[simp] lemma ncard_coe_finset (s : finset α) : (s : set α).ncard = s.card :=
+by rw [ncard_eq_to_finset_card, finset.finite_to_set_to_finset]
+
+lemma infinite.ncard (hs : s.infinite) : s.ncard = 0 :=
+@nat.card_eq_zero_of_infinite _ hs.to_subtype
+
+lemma ncard_univ (α : Type*) : (univ : set α).ncard = nat.card α :=
+begin
+  cases finite_or_infinite α with h h,
+  { haveI := @fintype.of_finite α h,
+    rw [ncard_eq_to_finset_card, finite.to_finset_univ, finset.card_univ,
+      nat.card_eq_fintype_card]},
+  rw [(@infinite_univ _ h).ncard, @nat.card_eq_zero_of_infinite _ h],
+end
+
+@[simp] lemma ncard_empty (α : Type*) : (∅ : set α).ncard = 0 := by simp only [ncard_eq_zero]
+
+lemma ncard_pos (hs : s.finite . to_finite_tac) : 0 < s.ncard ↔ s.nonempty :=
+by rw [pos_iff_ne_zero, ne.def, ncard_eq_zero hs, nonempty_iff_ne_empty]
+
+lemma ncard_ne_zero_of_mem (h : a ∈ s) (hs : s.finite . to_finite_tac) : s.ncard ≠ 0 :=
+((ncard_pos hs).mpr ⟨a,h⟩).ne.symm
+
+lemma finite_of_ncard_ne_zero (hs : s.ncard ≠ 0) : s.finite :=
+s.finite_or_infinite.elim id (λ h, (hs h.ncard).elim)
+
+lemma finite_of_ncard_pos (hs : 0 < s.ncard) : s.finite :=
+finite_of_ncard_ne_zero hs.ne.symm
+
+lemma nonempty_of_ncard_ne_zero (hs : s.ncard ≠ 0) : s.nonempty :=
+by {rw nonempty_iff_ne_empty, rintro rfl, simpa using hs}
+
+@[simp] lemma ncard_singleton (a : α) : ({a} : set α).ncard = 1 := by simp [ncard_eq_to_finset_card]
+
+lemma ncard_singleton_inter : ({a} ∩ s).ncard ≤ 1 :=
+begin
+  classical,
+  rw [←inter_self {a}, inter_assoc, ncard_eq_to_finset_card,
+    finite.to_finset_inter, finite.to_finset_singleton],
+  { apply finset.card_singleton_inter},
+  all_goals {apply to_finite},
+end
+
+section insert_erase
+
+@[simp] lemma ncard_insert_of_not_mem (h : a ∉ s) (hs : s.finite . to_finite_tac) :
+  (insert a s).ncard = s.ncard + 1 :=
+begin
+  classical,
+  haveI := hs.fintype,
+  rw [ncard_eq_to_finset_card, ncard_eq_to_finset_card, finite.to_finset_insert,
+    finset.card_insert_of_not_mem],
+  rwa [finite.mem_to_finset],
+end
+
+lemma ncard_insert_of_mem (h : a ∈ s) : ncard (insert a s) = s.ncard := by rw insert_eq_of_mem h
```

Comment:
```suggestion
lemma ncard_insert_of_mem (h : a ∈ s) : ncard (insert a s) = s.ncard :=
by rw insert_eq_of_mem h
```
We usually go to a newline unless the proof it's `rfl` (but this not important of course).